### DISCUSSION TITLE
Add source field to all open tables

### DISCUSSION THREAD

**Discussion by daviewales** [saulpw/visidata#2281] [Show and tell]
I had a bunch of tables I wanted to concat, but I didn't want to lose the source information for each.

I figured out how to do it with `exec-python`.

1. Start on the Sheets sheet, and select the sheets of interest.
2. Press `space`, then run the long-command `exec-python`
3. Press Ctrl+o to edit the Python expression in $EDITOR
4. Enter the following code, then save and exit:

``` python
for s in sheet.selectedRows:
    col = s.addColumn(ExprColumn('source', 'sheet.source'))
    s.freeze_col(col)
```

Now, all the individual sheets have a source column, which will be properly preserved when you concat them together.

(Should probably just add this to my `.visidatarc`... But cool we can do it ad-hoc.)

**Comment by daviewales** [comment]
Actually, this looked like it worked. But after I concat, they all have the same source. :(

**Reply by saulpw** [reply]
I think you almost have it!  Try reversing the order of addColumn and freeze_col:

```
s.addColumn(s.freeze_col(ExprColumn('source', 'sheet.source')))`
```

Since you want to add the frozen column and not the one that will change depending on what sheet it's on.

**Reply by daviewales** [reply]
Hmm... It's still not quite working for me.

If I do this:

``` python
for s in sheet.selectedRows:
    s.addColumn(s.freeze_col(ExprColumn('source', 'sheet.source')))
```

Then I get a new column called `source_frozen` with an error message: `use addColumn() on all columns!`:

``` python
Traceback (most recent call last):
File ".../visidata/lib/python3.10/site-packages/visidata/wrappers.py", line 105, in wrapply
return func(*args, **kwargs)
File ".../visidata/lib/python3.10/site-packages/visidata/column.py", line 340, in getValue
return self.calcValue(row)
File ".../visidata/lib/python3.10/site-packages/visidata/column.py", line 551, in calcValue
r = self.sheet.evalExpr(self.compiledExpr, row, col=self)
File ".../visidata/lib/python3.10/site-packages/visidata/utils.py", line 181, in __getattr__
raise Exception(self.__msg)
Exception: use addColumn() on all columns
```

I also tried adding the expr column, then freezing it:

``` python
for s in sheet.selectedRows:
    col = s.addColumn(ExprColumn('source', 'sheet.source'))
    s.addColumn(s.freeze_col(col))
```

This adds the expression column `source`, and the frozen column `source_frozen`.
But the frozen column only contains values for the first sheet after concatenating the sheets together.
It's `NULL` for the rest.

I've even tried opening all the sheets before running the above exec command, to ensure they are all loaded, and opening all the sheets after the above exec command, in case that helps, but it doesn't. (Using `gEnter`)

If I manually inspect each sheet after running the command above, then everything looks right.
But even after manually checking each sheet, when I concat, I only get source values for the first sheet, and `NULL` for the rest.

**Reply by daviewales** [reply]
Here's a command log:
[add-expr-col-on-selected-sheets.zip](https://github.com/saulpw/visidata/files/14118100/add-expr-col-on-selected-sheets.zip)
